Packs/PAN-OS/Integrations/Panorama/Panorama_test.py

```diff
@@ -1022,3 +1026,308 @@ def test_without_specify(self, requests_mock):
             requests_mock.get(Panorama.URL, text=data_file.read())
         Panorama.VSYS = None  # this a Panorama instance
         assert list(Panorama.devices()) == [('target1', 'vsys1'), ('target1', 'vsys2'), ('target2', None)]
+
+
+def load_xml_root_from_test_file(xml_file: str) -> ElementTree.Element:
+    """Given an XML file, loads it and returns the root element"""
+    xml_tree = ElementTree.parse(xml_file)
+    return xml_tree.getroot()
+
+
+MOCK_PANORAMA_SERIAL = "111222334455"
+MOCK_FIREWALL_1_SERIAL = "111111111111111"
+MOCK_FIREWALL_2_SERIAL = "222222222222222"
+MOCK_FIREWALL_3_SERIAL = "333333333333333"
+
+
+@pytest.fixture
+def mock_firewall():
+    mock_firewall = MagicMock(spec=Firewall)
+    mock_firewall.serial = MOCK_FIREWALL_1_SERIAL
+    mock_firewall.hostname = None
+
+    return mock_firewall
+
+
+@pytest.fixture
+def mock_panorama():
+    mock_panorama = MagicMock(spec=Panorama)
+    mock_panorama.serial = MOCK_PANORAMA_SERIAL
+    mock_panorama.hostname = None
+    return mock_panorama
+
+
+def mock_device_groups():
+    mock_device_group = MagicMock(spec=DeviceGroup)
+    mock_device_group.name = "test-dg"
+    return [mock_device_group]
+
+
+def mock_templates():
+    mock_template = MagicMock(spec=Template)
+    mock_template.name = "test-template"
+    return [mock_template]
+
+
+def mock_vsys():
+    mock_vsys = MagicMock(spec=Vsys)
+    mock_vsys.name = "vsys1"
+    return [mock_vsys]
+
+
+@pytest.fixture
+def mock_topology(mock_panorama, mock_firewall):
+    from Panorama import Topology
+    topology = Topology()
+    topology.panorama_objects = {
+        MOCK_PANORAMA_SERIAL: mock_panorama,
+    }
+    topology.firewall_objects = {
+        MOCK_FIREWALL_1_SERIAL: mock_firewall
+    }
+    topology.ha_active_devices = {
+        MOCK_PANORAMA_SERIAL: mock_panorama,
+        MOCK_FIREWALL_1_SERIAL: mock_firewall
+    }
+    return topology
+
+
+class TestTopology:
+    """Tests the Topology class and all of it's methods"""
+    SHOW_HA_STATE_ENABLED_XML = "test_data/show_ha_state_enabled.xml"
+    SHOW_HA_STATE_DISABLED_XML = "test_data/show_ha_state_disabled.xml"
+    SHOW_DEVICES_ALL_XML = "test_data/panorama_show_devices_all.xml"
+
+    @patch("Panorama.Topology.get_all_child_firewalls")
+    @patch("Panorama.run_op_command")
+    def test_add_firewall_device_object(self, patched_run_op_command, _, mock_firewall):
```

Comment:
@patch decorator passes the patched object as a function argument - if there's a way to prevent this behavior we can remove these _ arguments  in the test functions that use them.

Reply:
not sure how, need to check

Reply:
not sure how, need to check